Add canonical column to eth.header_cids

telackey opened the pull request:
Add boolean 'canonical' column to eth.header_cids.

At any given moment, there should only ever be one canonical block at a particular height.

telackey committed:
Add canonical column to eth.header_cids

telackey committed:
NOT NULL

roysc reviewed (changes requested):
Annoyingly, we'll also need to update the `TYPE header_result` which replicates `header_cids` for function interfaces in `00016_create_stored_procedures.sql`.

Also, from a design standpoint, I'm not sure about this, since canonicity isn't an intrinsic property of the block, and it means updating existing blocks on reorgs.

This is simple and performant (though the canonical check should never cause much bottleneck) but somewhat redundant when the vast majority of blocks are canonical anyway, and past a certain point we can safely ignore (or drop) non-canon blocks.

dboreham commented:
> it means updating existing blocks on reorgs

Isn't that the whole point? (or conversely : how would we store state about what's canonical without updating that state on reorg?)

telackey committed:
Update stored procedures for new schema.

telackey commented:
Additional context: 

https://github.com/cerc-io/ipld-eth-server/issues/251

telackey committed:
Switch to sql syntax, since it can be inlined.

telackey commented:
I updated `TYPE header_result` and I also rewrote the canonical_header_hash() function simply to use this column.  I also switched its language to `sql` since a `sql` function returning a query can be inlined by the query planner.  I haven't yet confirmed whether it _is_ inlined in our actual queries though.

I would like to drop the function entirely (see https://github.com/cerc-io/ipld-eth-server/issues/251) but this is a step in that direction.

telackey committed:
Fix indent

telackey committed:
Fix indent

roysc commented:
> Isn't that the whole point? (or conversely : how would we store state about what's canonical without updating that state on reorg?)

I mean that it doesn't have to be stored as part of the block. We could have a transient meta table keeping track of the non-canonical blocks from recent epochs. That would be append-only, so it would jibe with the copying-from-CSV pattern used in production.

dboreham commented:
> > Isn't that the whole point? (or conversely : how would we store state about what's canonical without updating that state on reorg?)
> 
> I mean that it doesn't have to be stored as part of the block. We could have a transient meta table keeping track of the non-canonical blocks from recent epochs. That would be append-only, so it would jibe with the copying-from-CSV pattern used in production.

Does this assume that we delete the non-canonical blocks when they fall below the finality threshold?

roysc commented:
> Does this assume that we delete the non-canonical blocks when they fall below the finality threshold?

Right, they could be pruned by some housekeeping service (maybe along with the blocks themselves)

dboreham commented:
> > Does this assume that we delete the non-canonical blocks when they fall below the finality threshold?
> 
> Right, they could be pruned by some housekeeping service (maybe along with the blocks themselves)

I think we need to retain non-canonical blocks from the pre-merge period because they could be useful for some analytics.
It feels safer to mark blocks as non-canonical vs deleting.

roysc commented:
> A secondary housekeeping service could not provide that consistency without increased difficulty and complexity.

To clarify, I'm referring to a service that garbage-collects the non-canonical markers from finalized epochs. There should be no ill effects from leaving them there, we would just want to keep that table tidy.

The transactional semantics would be essentially the same - when the new canon block is inserted, the ids of any displaced blocks are added to the meta table in the same tx. Any hashes present in the table are known as non-canonical.

i-norden reviewed (approved):
I'm of the (not super strong) opinion that marking the blocks canonical in-place is fine, particularly since we _do_ want to retain non-canonical blocks rather than garbage collect them below the finality threshold. Adding another table would require performing another join or query on said table in order to determine canonicity of the eth.header_cids content. If we _were_ using this meta data table to inform garbage collection of eth.header_cids, such that we could then blindly treat the eth.header_cids as a pure, canonical view I would be more in favor of this approach. 

The other issue is that we need to represent "canonicity" even in the non-final region of the chain, the latest two epochs. All we are really representing in this region is a consistent view of what geth considers the current working tip of the chain.

My only other comment was in regards to removing the canonical_header_hash function, but it looks like you already adjusted this to simply use this new field and are planning on removing it entirely after removing its usage in ipld-eth-server 👍
